### PR TITLE
Migrate from kapt to Anvil compiler api

### PR DESCRIPTION
**Pull request by changusmc**
This is a rewrite of the code generation of Kaiken from kapt to using the [Anvil compiler API](https://github.com/square/anvil/blob/main/compiler-api/README.md). The reason for using Anvil to generate code was because of its simplicity and it fits our needs.

In this rewrite, the interface is no longer being generated in Java, it will be generated in Kotlin. 

This is a backwards incompatible change to the consumer.

### CODE REVIEW

File: integration-tests/build.gradle [hunk 1]
```diff
@@ -25,6 +27,23 @@ dependencies {
 
     androidTestImplementation project(':testing')
 }
-kapt {
-    correctErrorTypes = true
+
+pluginManager.withPlugin('kotlin-kapt') {
+    kapt {
```
Comment (digitalbuddha) [inline integration-tests/build.gradle:32]: This should not be needed anymore. Corrsct error types was for Java generation AFAIK

File: processor/src/main/java/com/dropbox/kaiken/processor/InjectableActivityValidator.kt [hunk 2]
```diff
@@ -1,55 +1,53 @@
...
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.jetbrains.kotlin.resolve.descriptorUtil.getAllSuperClassifiers
+
+internal fun validateActivity(descriptor: ClassDescriptor?, clazz: KtClassOrObject) {
+    check(descriptor != null) {
+        "The ClassDescriptor for ${clazz.name} is null"
+    }
+
+    check(clazz.fqName != null) {
+        "The ClassDescriptor for ${clazz.name} is null"
```
Comment (digitalbuddha) [inline processor/src/main/java/com/dropbox/kaiken/processor/InjectableActivityValidator.kt:18]: Copy pasta this seems like wrong error message

File: processor/src/main/java/com/dropbox/kaiken/processor/InjectableActivityValidator.kt [hunk 3]
```diff
@@ -1,55 +1,53 @@
...
+    }
+
+    val fqName: FqName = requireNotNull(clazz.fqName) {
```
Comment (digitalbuddha) [inline processor/src/main/java/com/dropbox/kaiken/processor/InjectableActivityValidator.kt:21]: I'm not sure we need the check then a require not null. That should be same result

File: processor/src/main/java/com/dropbox/kaiken/processor/InjectableActivityValidator.kt [hunk 4]
```diff
@@ -1,55 +1,53 @@
...
+        }"
+    }
+
+    check(descriptor.implementsInjectorHolder()) {
+        "The class ${fqName.shortName()} does not implement" +
+            " DependencyProviderResolver"
     }
 }
+
+internal fun ClassifierDescriptor.isActivity(): Boolean =
```
Comment (digitalbuddha) [inline processor/src/main/java/com/dropbox/kaiken/processor/InjectableActivityValidator.kt:45]: Not call this isAndroidxActivty or actually check whether it's a component activity

File: integration-tests/build.gradle (hunk 1 above)
Comment (changusmc) [reply integration-tests/build.gradle:32]: Done

File: processor/src/main/java/com/dropbox/kaiken/processor/InjectableActivityValidator.kt (hunk 3 above)
Comment (changusmc) [reply processor/src/main/java/com/dropbox/kaiken/processor/InjectableActivityValidator.kt:21]: Done

File: processor/src/main/java/com/dropbox/kaiken/processor/InjectableActivityValidator.kt (hunk 4 above)
Comment (changusmc) [reply processor/src/main/java/com/dropbox/kaiken/processor/InjectableActivityValidator.kt:45]: I'll use similar naming as before https://github.com/dropbox/kaiken/blob/9a10c03f71d82fb783995308e96d5a620dc17b45/processor/src/main/java/com/dropbox/kaiken/processor/internal/ElementUtils.kt#L34

Comment (digitalbuddha) [reply processor/src/main/java/com/dropbox/kaiken/processor/InjectableActivityValidator.kt]: Hmm OK it was bad naming. Also I can't remember why we needed componentactivity but that was a recent change

**Review by rodgavila — APPROVED** [review]
Thanks @changusmc for the changes! Looks good!

**Comment by digitalbuddha** [comment]
Major Rev bump and release next?

**Comment by changusmc** [comment]
> Major Rev bump and release next?

yes sir! and dang it.. i think i did a merge and not squash merge

**Comment by digitalbuddha** [comment]
Good. I don't like squashes 😝

**Comment by vRallev** [comment]
👀 
image

**Pull request merged**